Storage/api.h

```diff
@@ -480,11 +553,119 @@ class RESTfulStorage {
+    data_->handler_routes_.emplace_back(path, route.resource_args_mask);
+    data_->handlers_scope_ += HTTP(data_->port_).Register(path, route.resource_args_mask, route.handler);
   }
 
+  template <typename T>
+  static std::shared_ptr<CurrentSuper> ParseCQRSRequest(const Request& request) {
+    // TODO(dkorolev): Proper exception types here!
+    auto object = std::make_shared<T>();
+    if (reflection::FieldCounter<T>::value > 0) {
+      if (!request.body.empty()) {
+        ParseJSON(request.body, *object);
+      } else {
+        request.url.query.FillObject(*object);
+      }
+    }
+    return object;
+  }
+
+  void RegisterCQRSHandlers(STORAGE_IMPL& storage, const std::string& restful_url_prefix) {
+    const Data& data = *data_;
+
+    const auto cqrs_query_handler = [&data, &storage, restful_url_prefix](Request request) {
+      if (request.url_path_args.empty()) {
+        request(Response(cqrs::CQRSHandlerNotSpecified(), HTTPResponseCode.NotFound));
+      } else {
+        std::lock_guard<std::mutex> lock(data.cqrs_handlers_mutex_);
+        const auto cit = data.cqrs_query_map_.find(request.url_path_args[0]);
+        if (cit != data.cqrs_query_map_.end()) {
+          const auto& f_parse_query_body = cit->second.first;
+          const auto& f_run_query = cit->second.second;
+          auto generic_input = RESTfulGenericInput<STORAGE_IMPL>(storage, restful_url_prefix);
+          using CQRSHandlerImpl = typename REST_IMPL::template RESTfulCQRSHandler<STORAGE_IMPL>;
+          CQRSHandlerImpl handler;
+          try {
+            std::shared_ptr<CurrentSuper> type_erased_query = f_parse_query_body(request);
+            handler.Enter(std::move(request),
+                          // Capture by reference since this lambda is run synchronously.
+                          [&handler, &f_run_query, &generic_input, &type_erased_query](Request request) {
+                            const STORAGE_IMPL& storage = generic_input.storage;
+                            storage.ReadOnlyTransaction(
+                                        // TODO(dkorolev): Lifetime management here, via Owner/Borrower.
+                                        // Capture local variables by value for safe async transactions.
+                                        [&storage, &f_run_query, handler, generic_input, type_erased_query](
+                                            immutable_fields_t fields) -> Response {
+                                          return f_run_query(
+                                              fields, std::move(type_erased_query), generic_input.restful_url_prefix);
+                                        },
+                                        std::move(request)).Detach();
+                          });
+          } catch (const Exception& e) {
+            // An exception has occurred while parsing user request.
+            request(Response("TODO(dkorolev): Error message here. Or reuse the `BadJSON` handler.",
```

Comment:
Sure, I understand that, though I think keeping the "public" part of it clean of "internals" such as TODOs is a good habit even in WIP code.